Mathlib/CategoryTheory/FiberedCategory/Cartesian.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2024 Calle Sönne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Lezeau, Calle Sönne
+-/
+
+import Mathlib.CategoryTheory.FiberedCategory.HomLift
+
+/-!
+# Cartesian morphisms
+
+This file defines cartesian resp. strongly cartesian morphisms in a based category.
+
+## Main definitions
+`IsCartesian p f φ` expresses that `φ` is a cartesian arrow lying over `f` with respect to `p` in
+the sense of SGA 1 VI 5.1. This means that for any morphism `φ' : a' ⟶ b` lying over `f` there is
+a unique morphism `τ : a' ⟶ a` lying over `𝟙 R`, such that `φ' = τ ≫ φ`.
+
+## References
+* [A. Grothendieck, M. Raynaud, *SGA 1*](https://arxiv.org/abs/math/0206203)
+-/
+
+universe v₁ v₂ u₁ u₂
+
+open CategoryTheory Functor Category IsHomLift
+
+namespace CategoryTheory
+
+variable {𝒮 : Type u₁} {𝒳 : Type u₂} [Category.{v₁} 𝒮] [Category.{v₂} 𝒳] (p : 𝒳 ⥤ 𝒮)
+
+/-- The proposition that a morphism `φ : a ⟶ b` in `𝒳` lying over `f : R ⟶ S` in `𝒮` is a
+cartesian arrow, see SGA 1 VI 5.1. -/
+class Functor.IsCartesian {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) extends
+    IsHomLift p f φ : Prop where
+  universal_property {a' : 𝒳} (φ' : a' ⟶ b) [IsHomLift p f φ'] :
+      ∃! χ : a' ⟶ a, IsHomLift p (𝟙 R) χ ∧ χ ≫ φ = φ'
+
+namespace IsCartesian
+
+variable {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) [IsCartesian p f φ]
+
+section
+
+variable {a' : 𝒳} (φ' : a' ⟶ b) [IsHomLift p f φ']
+
+/-- Given a cartesian arrow `φ : a ⟶ b` lying over `f : R ⟶ S` in `𝒳`, a morphism `φ' : a' ⟶ b`
+lifting `𝟙 R`, then `inducedMap f φ φ'` is the morphism `a' ⟶ a` obtained from the universal
+property of `φ`. -/
+noncomputable def inducedMap : a' ⟶ a :=
```

Comment:
I am not sure about the namespaces here. I would suggest placing `inducedMap` (and some other lemmas and definitions) in the `Functor` namespace, as `Functor.liftOfIsCartesian`. Below, the factorization could be `fac_of_isCartesian`. (This way, the names would kind of parallel the names in the limits api.)

Reply:
Good suggestion!

Reply:
I ended up trying to mirror the `IsLimit` names like your suggestion, but I'm not sure about the `Functor` namespace. Partly because this leads to especially long names later on like `liftOfIsStronglyCartesian`, and also because it doesn't feel like these are properties of the functor but rather the morphism (with respect to a given functor, of course). Also, like I said in the comment above, I'm not sure if Cartesian arrows should be directly formulated in terms of functors, or if they should be formulated in terms of `BasedCategory` as in another one of my PRs.

Reply:
Another problem is that "cartesian" is a terminology that is used beyond the case of fibered categories. Then, I think that `Functor` is better.

Reply:
Okay that's a good point. Is it maybe possible to put it in a namespace `Functor.IsCartesian`, so that these short names can be used? I just feel it will be annoying to write `ofIsStronglyCartesian` everywhere (if I make the same changes in the following PR).